Blog post about 1.7 release (#192)
kubewarden/kubewarden.io: content/blog/2023/09/kubewarden-1.7-release.md

---
title: Kubewarden 1.7.0 release
authors:
  - Flavio Castelli
date: 2023-09-12
---

> Today we are pleased to announce the release of Kubewarden 1.7.0! 🎊 🥳

Reviewer: ```suggestion
Today we are delighted to announce the release of Kubewarden 1.7.0! 🎊 🥳
```

Aside from the bug and stability fixes, this release is packed with new features.
This post highlights the main changes, detailed blog entries will come in the next weeks covering each feature in depth.

> ## Audit Scanner

Reviewer: ```suggestion
## Audit scanner
```

A new component has been added to the Kubewarden stack. Its name is `audit-scanner` and
it allows administrators to assess the compliance level of the clusters secured by Kubewarden.

> At regular intervals, or on demand, the audit scanner inspects all the resources defined
> inside of your Kubernetes cluster and identifies the ones that are not compliant with the
> policies that are currently enforced.

Reviewer: ```suggestion
At regular intervals, or on demand, the audit scanner inspects all the resources defined inside your Kubernetes cluster. 
It identifies those not compliant with the policies that are currently enforced.
```

> Why is this useful? The number, types and configurations of the policies enforced on a cluster
> vary over its lifetime. A resource that was considered compliant some weeks ago, today
> might not be valid anymore since the introduction of a new policy.

Reviewer: ```suggestion
Why is this useful? The number, types and configurations of the policies enforced on a cluster
vary over its lifetime. 
A resource that was considered compliant recently may no longer be, since the introduction of a new policy.
```

> The audit scanner writes its findings using two new types of Kubernetes Custom Resources:
> `PolicyReport` and `ClusterPolicyReport`. Each Kubernetes Namespace inspected by the Audit Scanner
> will feature one `PolicyReport` object that contains the results of the audit.
> For cluster-wide resources (like Namespace, PersistentVolume, &hellip;), the Audit Scanner
> creates one `ClusterPolicyReport` object per cluster.

Reviewer: ```suggestion
The audit scanner writes its findings using two new types of Kubernetes Custom Resources:
`PolicyReport` and `ClusterPolicyReport`. Each Kubernetes Namespace inspected by the audit scanner
will feature one `PolicyReport` object that contains the results of the audit.
For cluster-wide resources (like Namespace, PersistentVolume, &hellip;), the audit scanner
creates one `ClusterPolicyReport` object per cluster.
```

We will publish a dedicated blog post about this new feature. In the meantime, you

> can learn more about it by reading [this section](https://docs.kubewarden.io/explanations/audit-scanner)
> of our docs.

Reviewer: ```suggestion
can learn more about it by reading [this section](https://docs.kubewarden.io/explanations/audit-scanner).
```

## New Policy Type

Kubewarden policies can be written using either a [traditional programming language](https://docs.kubewarden.io/writing-policies) (like Go, Rust, C#, Swift, &hellip;) or using a domain-specific language like [Rego](https://docs.kubewarden.io/writing-policies/rego/intro-rego).

> It is required that the programming language can generate the necessary WebAssembly module for use by KubeWarden.

Reviewer 2: KubeWarden or Kubewarden?

Author: good catch, it's `Kubewarden`

However, it fails at compiling complex Go codebases. This isn't usually a problem for the average Kubewarden policy,
but we had something in mind that exceeded TinyGo's capabilities.

Last month (Aug 2023) Go 1.21 was released. With this release, the official Go compiler can
target also the [WebAssembly System Interface](https://wasi.dev/) (aka _"WASI"_).
This means it's now possible to create WebAssembly modules that can be evaluated outside of a browser.

The Go WASI support is pretty good, but there's still [one limitation](https://github.com/golang/go/issues/42372)
that prevents the creation of regular Kubewarden policies.
To circumvent this limitation, and be able to leverage the official Go compiler, we defined a type of Kubewarden policy that we
called _"vanilla WASI"_.

This new type of policy should be used only in special circumstances because it has its own set of limitations.
You can read more about this type of policy inside of [our docs](https://docs.kubewarden.io/writing-policies/wasi).
A detailed blog post about this new type of policy will be published in the next weeks.

## Experiment: reuse Kyverno DSL

> As written in the previous section, we have been playing with the WASI support
> introduced by the Go 1.21 release. We did that to create a Kubewarden policy capable of interpreting the Kyverno DSL,

Reviewer: ```suggestion
It's clear from the previous section, we have been playing with the WASI support
introduced by the Go 1.21 release. We did this to create a Kubewarden policy capable of interpreting the Kyverno DSL,
```

this required the full power of the official Go compiler.

You can find this policy [here](https://github.com/kubewarden/kyverno-dsl-policy). The code is in its early days,
but it's already able to reuse certain types of Kyverno policies.

The policy is highly experimental and has limitations.
We would like to know what the community thinks about this idea. Should we keep
pushing this concept forward, are you interested in reusing the Kyverno DSL?

## Go policies improvements

> Going back to the regular Kubewarden policies written in Go and compiled using TinyGo, we have

Reviewer: ```suggestion
Returning to the regular Kubewarden policies written in Go and compiled using TinyGo, we have
```

good news for policy authors.

> The latest release of TinyGo (v0.29.0 published in August too), provides great improvements. The biggest one

Reviewer: ```suggestion
The latest release of TinyGo (v0.29.0 published in August too), provides some great enhancements. The biggest one
```

> is the support of the [`reflection`](https://pkg.go.dev/reflect) module. Thanks to this change it is

Reviewer: ```suggestion
is the support of the [`reflection`](https://pkg.go.dev/reflect) module. Thanks to this change it's
```

now possible to use the "[`encoding/json`](https://pkg.go.dev/encoding/json)" module.

> Before this change policy authors had to resort to using the [`easyjson`](https://github.com/mailru/easyjson/)
> library to handle the marshalling and unmarshalling. While this worked, the developer experience was impacted
> by this workaround.

Reviewer: ```suggestion
Before this change policy authors had to resort to using the [`easyjson`](https://github.com/mailru/easyjson/)
library to handle marshalling and unmarshalling. It worked but the developer experience wasn't ideal with this workaround.
```
